source/iopipe/json/formatter.d

```diff
@@ -0,0 +1,641 @@
+        state = State.First;
+
+        ++spacing;
+        putStr("[");
+        putIndent();
+    }
+
+    void endArray() {
+        if (!canCloseArray())
+            throw new JSONIopipeException("endArray not allowed in current state");
+
+        --spacing;
+        putIndent();
+        putStr("]");
+
+        popContainer();
+    }
+
+    // call before each element
+    void beginArrayValue()
+    {
+        if (!canAddArrayMember())
+            throw new JSONIopipeException("beginArrayValue not allowed in current state");
+    }
+
+    void addMember(T)(T key) {
+        addMember(key, memberOptions);
+    }
+
+    // First/Member -> Value(before string key) -> Member(after string key) -> Value(before value) -> Member(after value) ...
+    void addMember(T)(T key, MemberOptions options) {
+        
+        if (!canAddObjMember())
+            throw new JSONIopipeException("addMember not allowed in current state");
+
+        // prepare for a key string
+        state = State.Value;
+
+        static if (is(typeof(key) == string))
+        {
+            if (options.escapeKey)
+            {
+                beginString(options.quoteChar);
+                addStringData(key);
+                endString();
+            }
+            else
+            {
+                beginString(options.quoteChar);
+                addStringData!(false, false)(key);
+                endString();
+            }
+        }
+        else
+        {
+            putStr(to!string(key));
+        }
+
+        final switch (options.colonSpacing)
+        {
+        case MemberOptions.ColonSpacing.none:
+            putStr(":");
+            break;
+        case MemberOptions.ColonSpacing.after:
+            putStr(": ");
+            break;
+        case MemberOptions.ColonSpacing.both:
+            putStr(" : ");
+            break;
+        }
+
+        // prepare for a key's value
+        state = State.Value;
+    }
+
+    void beginString(char quoteChar = '"') {
+        if (!canAddValue())
+            throw new JSONIopipeException("beginString not allowed in current state");
+        
+        if (currentQuoteChar != char.init)
+          throw new JSONIopipeException("nested beginString not allowed");
+
+        if (quoteChar != '"' && quoteChar != '\'')
+            throw new JSONIopipeException("Invalid quote character");
+
+        currentQuoteChar = quoteChar;
+        putStr((&quoteChar)[0 .. 1]);
+    }
+
+    // will automatically escape string data as needed.
+    void addStringData(bool validate = true, bool addEscapes = true, T)(T value) {
+        if (currentQuoteChar == char.init)
+            throw new JSONIopipeException("cannot add string data outside of beginString/endString");
+
+        static if (validate) {
+            // Validate that the string doesn't contain invalid characters
+            foreach(char c; value) {
+                if (c < 0x20 && c != '\t' && c != '\n' && c != '\r') {
```

Comment:
This isn't exactly correct, it needs to check for quote characters and lone backslashes.

In addition, I would expect both add escapes and validate as true should work with characters like newline, which should replace with a literal `\n`

I know I made this API in the discussion, I think it's wrong now. You can validate, add escapes, or do neither. You can't do both (adding escapes would ensure a valid string).

Maybe change to an enum?

passThru -> don't do anything, assume the string data is correct
addEscapes -> any invalid character is escaped into a valid sequence (this should be the default)
validate -> verify escapes are correctly used in the data.